### PR TITLE
[GPU] Support FC+eltwise fusion in fp16 for OneDNN

### PR DESCRIPTION
**Pull request by hyunback**
- Enable oneDNN FC and eltwise fusion in fp16

### Tickets:
 - *101693*

### CODE REVIEW

File: src/plugins/intel_gpu/tests/fusions/fully_connected_fusion_test.cpp [hunk 1]
```diff
@@ -146,6 +156,13 @@ class FullyConnectedFusingTestOneDNN : public BaseFusingTest<fully_connected_tes
...
 
+#define CASE_FC_FP16_1 { 1, 3 }, { 1, 4 }, { 4, 3 }, data_types::f16, format::bfyx, data_types::f16, format::oiyx, data_types::f32, format::bfyx
+#define CASE_FC_FP16_2 { 2, 3 }, { 2, 4 }, { 4, 3 }, data_types::f16, format::bfyx, data_types::f16, format::oiyx, data_types::f32, format::bfyx
+#define CASE_FC_FP16_3 { 2, 32 }, { 2, 16 }, { 16, 32 }, data_types::f16, format::bfyx, data_types::f16, format::oiyx, data_types::f32, format::bfyx
+#define CASE_FC_FP16_3D_1 { 2, 32, 3 }, { 2, 32, 16 }, { 16, 3, 1 }, data_types::f16, format::bfyx, data_types::f16, format::oiyx, data_types::f32, format::bfyx
+#define CASE_FC_FP16_3D_2 { 1, 1, 3 }, { 1, 1, 32 }, { 32, 3, 1 }, data_types::f16, format::bfyx, data_types::f16, format::oiyx, data_types::f32, format::bfyx
+// #define CASE_FC_FP16_3D_3 { 2, 3, 1 }, { 2, 3, 15 }, { 15, 1, 1 }, data_types::f16, format::bfyx, data_types::f16, format::oiyx, data_types::f32, format::bfyx
+// #define CASE_FC_FP16_3D_4 { 2, 1, 1 }, { 2, 1, 15 }, { 15, 1, 1 }, data_types::f16, format::bfyx, data_types::f16, format::oiyx, data_types::f32, format::bfyx
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/fusions/fully_connected_fusion_test.cpp:165]: why is it commented out? If it is necessary but failing, please leave a comment. If it is not necessary, please remove it.

File: src/plugins/intel_gpu/tests/fusions/fully_connected_fusion_test.cpp [hunk 2]
```diff
@@ -81,14 +82,23 @@ class FullyConnectedFusingTestOneDNN : public BaseFusingTest<fully_connected_tes
...
         for (auto& forcing : impl_forcing)
             if (forcing.first == "fc_prim")
                 forcing_format = forcing.second.output_format;
 
-        ov::intel_gpu::ImplementationDesc conv_impl = { forcing_format, "", impl_types::onednn };
-        cfg_fused.set_property(ov::intel_gpu::force_implementations(ov::intel_gpu::ImplForcingMap{ { "fc_prim", conv_impl } }));
+        ov::intel_gpu::ImplementationDesc fc_impl = { forcing_format, "", impl_types::onednn };
+        cfg_fused.set_property(ov::intel_gpu::force_implementations(ov::intel_gpu::ImplForcingMap{ { "fc_prim", fc_impl } }));
+
+        if (!p.ocl_kernel_name.empty()) {
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/fusions/fully_connected_fusion_test.cpp:93]: is `ocl_kernel_name` required? what about just forcing ref kernel for unfused case?

**Review by isanghao — CHANGES_REQUESTED** [review]
looks good in general. Let me check the performance.

File: src/plugins/intel_gpu/tests/fusions/fully_connected_fusion_test.cpp (hunk 2 above)
Comment (hyunback) [reply src/plugins/intel_gpu/tests/fusions/fully_connected_fusion_test.cpp:93]: If use force_implemention keyword, network always run with optimization path(prepare_qunatization, primitive_fusing, reorder_input...).
The purpose of fusing test is comparing non_fused vs fused network, but we can't guarantee pure none fused network if use force_implementaton in non_fused network.
FC and eltwise fusing in fp16 is currently blocked in clDNN.  So this cases is okay whether non_fused network use force_impl or not, but other tests used FullyconnectedFusingTestOneDNN have the issue for ref value correctness.
* Solution for removing ocl_kernel_name
1. Find and fix the high priority kernel(fully_connected_gpu_bs_f_bsv16_b1) failed reason and then run the high priority kernel without ocl_kernel_name force_implementation.
2. Make another FullyConnectedFusingTestOneDNN class only for FC and eltwise fusion. It always select ref kernel.
3. Nothing action

File: src/plugins/intel_gpu/tests/fusions/fully_connected_fusion_test.cpp (hunk 1 above)
Comment (hyunback) [reply src/plugins/intel_gpu/tests/fusions/fully_connected_fusion_test.cpp:165]: I'll remove it

File: src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp [hunk 3]
```diff
@@ -1008,6 +1008,29 @@ void reorder_inputs::run(program& p, layout_optimizer& lo, reorder_factory& rf)
...
+            for (const auto& fused_prim : node->get_fused_primitives()) {
+                if (fused_prim.is_type<eltwise>() &&
+                    one_of(fused_prim.typed_desc<eltwise>()->mode, {eltwise_mode::sum, eltwise_mode::sub, eltwise_mode::prod})) {
+                    auto fc_layout = node->get_output_layout();
+                    auto& data = node->get_dependency(fused_prim.dep_start_idx);
+                    auto data_layout = data.get_output_layout();
+
+                    if ((fc_layout.batch() == 1 || fc_layout.feature() == 1) ||
+                        (data_layout.batch() == 1 && data_layout.feature() == 1) ||
+                        (fc_layout.count() == data_layout.count())) {
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp:1026]: (maybe in a separate PR) could you add a comment on this condition? I just made a table and this condition looks good.

image

**Review by isanghao — APPROVED** [review]

File: src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp (hunk 3 above)
Comment (hyunback) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp:1026]: Okay I see, let's me add it next my PR

**Pull request merged**